File: frontend/src/lib/bookend.ts
/**
 * Project-level intro/outro bookend packaging (preview + config).
 * Mirrors pixelle_video/utils/bookend.py defaults.
 */

export interface BookendConfig {
  enabled: boolean;
  introSeconds: number;
  outroSeconds: number;
  introFadeSeconds: number;
  outroFadeSeconds: number;
}

export const DEFAULT_BOOKEND: BookendConfig = {
  enabled: true,
  introSeconds: 1.2,
  outroSeconds: 2.0,
  introFadeSeconds: 0.6,
  outroFadeSeconds: 1.0,
};

const clamp = (value: number, min: number, max: number) =>
  Math.min(max, Math.max(min, value));

export function normalizeBookendConfig(
  config?: Record<string, unknown> | null,
): BookendConfig {
  const cfg = config || {};
  const nested =
    cfg.bookend && typeof cfg.bookend === "object"
      ? (cfg.bookend as Record<string, unknown>)
      : {};
  const source: Record<string, unknown> = { ...DEFAULT_BOOKEND, ...nested, ...cfg };

  const enabledRaw = source.bookendEnabled ?? source.bookend_enabled ?? source.enabled ?? true;
  const enabled =
    typeof enabledRaw === "string"
      ? !["0", "false", "no", "off", ""].includes(enabledRaw.trim().toLowerCase())
      : Boolean(enabledRaw);

  let introSeconds = clamp(
    Number(source.bookendIntroSeconds ?? source.intro_seconds ?? source.introSeconds ?? DEFAULT_BOOKEND.introSeconds) || 0,
    0,
    5,
  );
  let outroSeconds = clamp(
    Number(source.bookendOutroSeconds ?? source.outro_seconds ?? source.outroSeconds ?? DEFAULT_BOOKEND.outroSeconds) || 0,
    0,
    6,
  );
  let introFadeSeconds = clamp(
    Number(
      source.bookendIntroFadeSeconds
        ?? source.intro_fade_seconds
        ?? source.introFadeSeconds
        ?? DEFAULT_BOOKEND.introFadeSeconds,
    ) || 0,
    0,
    3,
  );
  let outroFadeSeconds = clamp(
    Number(
      source.bookendOutroFadeSeconds
        ?? source.outro_fade_seconds
        ?? source.outroFadeSeconds
        ?? DEFAULT_BOOKEND.outroFadeSeconds,
    ) || 0,
    0,
    4,
  );

  if (!enabled) {
    introSeconds = 0;
    outroSeconds = 0;
    introFadeSeconds = 0;
    outroFadeSeconds = 0;
  } else {
    introFadeSeconds = introSeconds > 0 ? Math.min(introFadeSeconds, introSeconds) : 0;
    outroFadeSeconds = outroSeconds > 0 ? Math.min(outroFadeSeconds, outroSeconds) : 0;
  }

  return {
    enabled: enabled && (introSeconds > 0 || outroSeconds > 0),
    introSeconds,
    outroSeconds,
    introFadeSeconds,
    outroFadeSeconds,
  };
}

export type BookendPhase = "intro" | "content" | "outro";

export interface BookendPlaybackState {
  phase: BookendPhase;
  /** Time within content timeline (0-based). Negative in intro; > contentDuration in outro. */
  contentTime: number;
  /** 0–1 opacity for stage picture fade approximation */
  pictureOpacity: number;
  /** 0–1 multiplier for BGM volume */
  bgmGain: number;
  /** Whether narration should play */
  narrationEnabled: boolean;
}

/**
 * Map absolute playback clock (includes intro/outro) to bookend visual/audio state.
 */
export function getBookendPlaybackState(
  playbackTime: number,
  contentDuration: number,
  bookend: BookendConfig,
): BookendPlaybackState {
  const intro = bookend.enabled ? bookend.introSeconds : 0;
  const outro = bookend.enabled ? bookend.outroSeconds : 0;
  const fadeIn = bookend.enabled ? bookend.introFadeSeconds : 0;
  const fadeOut = bookend.enabled ? bookend.outroFadeSeconds : 0;
  const total = Math.max(0, contentDuration) + intro + outro;
  const t = Math.max(0, Math.min(total, playbackTime));

  if (intro > 0 && t < intro) {
    const local = t;
    const opacity = fadeIn > 0 ? Math.min(1, local / fadeIn) : 1;
    const bgmGain = fadeIn > 0 ? Math.min(1, local / fadeIn) : 1;
    return {
      phase: "intro",
      contentTime: local - intro,
      pictureOpacity: opacity,
      bgmGain,
      narrationEnabled: false,
    };
  }

  const contentEnd = intro + Math.max(0, contentDuration);
  if (outro > 0 && t >= contentEnd) {
    const intoOutro = t - contentEnd;
    const remaining = Math.max(0, outro - intoOutro);
    // Fade only in the last fadeOut seconds of the full timeline
    let opacity = 1;
    let bgmGain = 1;
    if (fadeOut > 0) {
      const fadeStart = total - fadeOut;
      if (t >= fadeStart) {
        const u = (t - fadeStart) / fadeOut;
        opacity = Math.max(0, 1 - u);
        bgmGain = Math.max(0, 1 - u);
      }
    }
    return {
      phase: "outro",
      contentTime: Math.max(0, contentDuration) + intoOutro,
      pictureOpacity: opacity,
      bgmGain,
      narrationEnabled: false,
      // remaining unused but kept for future UI
    };
  }

  // Content region — optional mild edge fades already applied outside
  void fadeIn;
  return {
    phase: "content",
    contentTime: t - intro,
    pictureOpacity: 1,
    bgmGain: 1,
    narrationEnabled: true,
  };
}

export function getPlaybackTotalDuration(contentDuration: number, bookend: BookendConfig): number {
  if (!bookend.enabled) return Math.max(0, contentDuration);
  return Math.max(0, contentDuration) + bookend.introSeconds + bookend.outroSeconds;
}

File: pixelle_video/utils/bookend.py
"""Project-level intro/outro bookend packaging defaults and normalization."""

from __future__ import annotations

from typing import Any


# Product defaults (see design: 片头 1.2s / 片尾 2.0s, fades nested inside).
DEFAULT_BOOKEND = {
    "enabled": True,
    "intro_seconds": 1.2,
    "outro_seconds": 2.0,
    "intro_fade_seconds": 0.6,
    "outro_fade_seconds": 1.0,
}


def _to_float(value: Any, default: float) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return default
    if number != number:  # NaN
        return default
    return number


def _clamp(value: float, minimum: float, maximum: float) -> float:
    return max(minimum, min(maximum, value))


def normalize_bookend_config(config: dict[str, Any] | None = None) -> dict[str, Any]:
    """
    Normalize project/export config into a canonical bookend dict.

    Accepts camelCase (frontend) and snake_case (backend) keys.
    Outro is additive with per-scene manual hold (hold stays on scenes; outro is extra).
    """
    cfg = config or {}

    if "bookendEnabled" in cfg or "bookend_enabled" in cfg:
        enabled_raw = cfg.get("bookendEnabled", cfg.get("bookend_enabled"))
    else:
        # Nested object support: config.bookend = { enabled, ... }
        nested = cfg.get("bookend") if isinstance(cfg.get("bookend"), dict) else {}
        enabled_raw = nested.get("enabled", DEFAULT_BOOKEND["enabled"])
        cfg = {**nested, **cfg}

    if isinstance(enabled_raw, str):
        enabled = enabled_raw.strip().lower() not in {"0", "false", "no", "off", ""}
    else:
        enabled = bool(enabled_raw) if enabled_raw is not None else True

    intro = _clamp(
        _to_float(
            cfg.get("bookendIntroSeconds", cfg.get("intro_seconds", cfg.get("introSeconds"))),
            DEFAULT_BOOKEND["intro_seconds"],
        ),
        0.0,
        5.0,
    )
    outro = _clamp(
        _to_float(
            cfg.get("bookendOutroSeconds", cfg.get("outro_seconds", cfg.get("outroSeconds"))),
            DEFAULT_BOOKEND["outro_seconds"],
        ),
        0.0,
        6.0,
    )
    intro_fade = _clamp(
        _to_float(
            cfg.get(
                "bookendIntroFadeSeconds",
                cfg.get("intro_fade_seconds", cfg.get("introFadeSeconds")),
            ),
            DEFAULT_BOOKEND["intro_fade_seconds"],
        ),
        0.0,
        3.0,
    )
    outro_fade = _clamp(
        _to_float(
            cfg.get(
                "bookendOutroFadeSeconds",
                cfg.get("outro_fade_seconds", cfg.get("outroFadeSeconds")),
            ),
            DEFAULT_BOOKEND["outro_fade_seconds"],
        ),
        0.0,
        4.0,
    )

    # Fades cannot exceed pad durations.
    if intro > 0:
        intro_fade = min(intro_fade, intro)
    else:
        intro_fade = 0.0
    if outro > 0:
        outro_fade = min(outro_fade, outro)
    else:
        outro_fade = 0.0

    if not enabled:
        intro = outro = intro_fade = outro_fade = 0.0

    return {
        "enabled": enabled and (intro > 0 or outro > 0),
        "intro_seconds": round(intro, 3),
        "outro_seconds": round(outro, 3),
        "intro_fade_seconds": round(intro_fade, 3),
        "outro_fade_seconds": round(outro_fade, 3),
    }
